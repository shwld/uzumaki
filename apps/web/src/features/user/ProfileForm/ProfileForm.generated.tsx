import * as Types from '../../../graphql/generated/graphql';

import gql from 'graphql-tag';
import * as Urql from 'urql';
export type DateString = string;
export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
export type ProfileForm_UpdateResultFragment = {
  __typename?: 'UserProfile';
  id: string;
  name: string;
  avatarImageUrl: string;
};

export type ProfileForm_UserProfileQueryVariables = Types.Exact<{
  [key: string]: never;
}>;

export type ProfileForm_UserProfileQuery = {
  __typename?: 'Query';
  viewer?:
    | {
        __typename?: 'Viewer';
        id: string;
        profile: {
          __typename?: 'UserProfile';
          id: string;
          name: string;
          avatarImageUrl: string;
        };
      }
    | undefined;
};

export type ProfileForm_UpdateUserProfileMutationVariables = Types.Exact<{
  input: Types.UpdateUserProfileInput;
}>;

export type ProfileForm_UpdateUserProfileMutation = {
  __typename?: 'Mutation';
  updateUserProfile:
    | { __typename?: 'InternalErrorResult' }
    | { __typename?: 'InvalidArgumentsResult' }
    | { __typename?: 'UnauthorizedResult' }
    | {
        __typename?: 'UpdateUserProfileSuccessResult';
        result: {
          __typename?: 'UserProfile';
          id: string;
          name: string;
          avatarImageUrl: string;
        };
      };
};

export const ProfileForm_UpdateResultFragmentDoc = gql`
  fragment ProfileForm_UpdateResult on UserProfile {
    id
    name
    avatarImageUrl
  }
`;
export const ProfileForm_UserProfileDocument = gql`
  query ProfileForm_UserProfile {
    viewer {
      id
      profile {
        id
        name
        avatarImageUrl
      }
    }
  }
`;

export function useProfileForm_UserProfileQuery(
  options?: Omit<
    Urql.UseQueryArgs<ProfileForm_UserProfileQueryVariables>,
    'query'
  >
) {
  return Urql.useQuery<
    ProfileForm_UserProfileQuery,
    ProfileForm_UserProfileQueryVariables
  >({ query: ProfileForm_UserProfileDocument, ...options });
}
export const ProfileForm_UpdateUserProfileDocument = gql`
  mutation ProfileForm_UpdateUserProfile($input: UpdateUserProfileInput!) {
    updateUserProfile(input: $input) {
      ... on UpdateUserProfileSuccessResult {
        result {
          ...ProfileForm_UpdateResult
        }
      }
    }
  }
  ${ProfileForm_UpdateResultFragmentDoc}
`;

export function useProfileForm_UpdateUserProfileMutation() {
  return Urql.useMutation<
    ProfileForm_UpdateUserProfileMutation,
    ProfileForm_UpdateUserProfileMutationVariables
  >(ProfileForm_UpdateUserProfileDocument);
}
