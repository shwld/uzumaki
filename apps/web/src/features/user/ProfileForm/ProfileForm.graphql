fragment ProfileForm_UpdateResult on UserProfile {
  id
  name
  avatarImageUrl
}

mutation ProfileForm_UpdateUserProfle($input: UpdateUserProfileInput!) {
  updateUserProfile(input: $input) {
    ... on UpdateUserProfileSuccessResult {
      result {
        ...ProfileForm_UpdateResult
      }
    }
  }
}
