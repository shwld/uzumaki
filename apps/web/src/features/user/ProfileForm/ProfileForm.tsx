import {
  Avatar,
  Box,
  Button,
  FormControl,
  FormErrorMessage,
  FormLabel,
  HStack,
  Input,
  Text,
  useId,
} from '@chakra-ui/react';
import { zodResolver } from '@hookform/resolvers/zod';
import { Scalars } from 'graphql-resolvers/src/generated/resolvers-types';
import { updateUserProfileArgsValidationSchema } from 'graphql-resolvers/src/modules/user-profile/mutation-resolvers/user-profile.update/update-user-profile-validation';
import { FC, useRef } from 'react';
import { useForm } from 'react-hook-form';
import {
  useProfileForm_UpdateUserProfileMutation,
  useProfileForm_UserProfileQuery,
} from './ProfileForm.generated';

type ProfileInput = {
  name: Scalars['String'];
  // avatarImageUrl: Scalars['String'];
};

export const ProfileForm: FC = () => {
  const [{ data }] = useProfileForm_UserProfileQuery();
  if (data?.viewer?.profile == null) return <></>;

  return (
    <ProfileEditForm
      defaultValues={data?.viewer?.profile}
      avatarImageUrl={data.viewer.profile.avatarImageUrl}
    />
  );
};

const ProfileEditForm: FC<{
  defaultValues: ProfileInput;
  avatarImageUrl?: string;
}> = ({ defaultValues, avatarImageUrl }) => {
  const id = useId();
  const fileId = useId();
  const file = useRef<HTMLInputElement>(null);
  const [result, updateUserProfile] =
    useProfileForm_UpdateUserProfileMutation();
  const {
    handleSubmit,
    register,
    formState: { errors },
  } = useForm<ProfileInput>({
    resolver: zodResolver(updateUserProfileArgsValidationSchema.shape.input),
    defaultValues,
  });
  const submit = handleSubmit(async input => {
    const image = file.current?.files?.[0];
    await updateUserProfile({
      input: {
        name: input.name,
        avatarImage: image,
      },
    });
  });

  return (
    <Box p={3} bg="gray.100">
      <Text>User Profile</Text>
      <form onSubmit={submit}>
        <HStack w="full">
          <FormControl isInvalid={errors.name != null}>
            <FormLabel htmlFor={id} mb={0}>
              Name
            </FormLabel>
            <Input type="text" id={id} {...register('name')} bgColor="white" />
            {errors.name && (
              <FormErrorMessage>{errors.name.message}</FormErrorMessage>
            )}
          </FormControl>
        </HStack>
        <HStack w="full">
          <FormControl>
            <FormLabel htmlFor={fileId} mb={0}>
              Avatar
              <Avatar src={avatarImageUrl} />
            </FormLabel>
            <input
              type="file"
              id={fileId}
              ref={file}
              style={{ display: 'none' }}
            />
          </FormControl>
        </HStack>
        <Button type="submit" colorScheme="blue" mt="3">
          Submit
        </Button>
      </form>
    </Box>
  );
};
