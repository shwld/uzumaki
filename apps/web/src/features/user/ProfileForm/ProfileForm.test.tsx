import { ProfileForm } from './ProfileForm';
import { render } from '@testing-library/react';
import { MockedUrqlProvider } from '~/test/MockedUrqlProvider';

describe('ProfileForm', () => {
  const renderComponent = () => {
    const renderResult = render(
      <MockedUrqlProvider>
        <ProfileForm />
      </MockedUrqlProvider>
    );
    return renderResult;
  };
  test('Snapshot', () => {
    expect(renderComponent().asFragment()).toMatchSnapshot();
  });
  test('success', () => {
    const { getByText } = renderComponent();
    expect(getByText('User Profile')).toBeTruthy();
  });
});
