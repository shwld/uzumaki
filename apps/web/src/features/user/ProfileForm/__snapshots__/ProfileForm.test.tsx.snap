// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileForm Snapshot 1`] = `
<DocumentFragment>
  <div>
    test
  </div>
</DocumentFragment>
`;
